racket-cas/racket-cas.rkt

```diff
@@ -1348,36 +1370,16 @@
 
   ; Implementation note: If u is normalized, then so will a partial sum,
   ; if we keep the original order.  
-  (define (reverse-terms->sum vs)    
-    (match vs
-      ['()      0]
-      [(list v) v]
-      [_        (cons '+ (reverse vs))])) ; a partial sum is normalized
-  
   (match u
-    ; a single imaginary term
-    ['@i                                  (values 0 1)]
-    [(list  '*   '@i u)                   (values 0 u)]
-    [(list* '*   '@i us)                  (values 0 (list* '* us))]
-    [(list  '* r '@i)    #:when (real? r) (values 0 r)]
-    [(list  '* r '@i u)  #:when (real? r) (values 0 `(* ,r ,u))]
-    [(list* '* r '@i us) #:when (real? r) (values 0 (list* '* r us))]
-    ; a sum of terms
-    [(list* '+ us)     
-     (let loop ([us us] [real '()] [imag '()])
+    [(Sum us)
        (match us
-         ['()         (values (reverse-terms->sum real)
-                              (reverse-terms->sum imag))]         
-         [(cons v vs) (match v
-                        ['@i                                  (loop vs real (cons 1               imag))]
-                        [(list  '*   '@i u)                   (loop vs real (cons u               imag))]
-                        [(list* '*   '@i us)                  (loop vs real (cons (list* '* us)   imag))]
-                        [(list  '* r '@i)    #:when (real? r) (loop vs real (cons r               imag))]
-                        [(list  '* r '@i u)  #:when (real? r) (loop vs real (cons `(* ,r ,u)      imag))]
-                        [(list* '* r '@i us) #:when (real? r) (loop vs real (cons (list* '* r us) imag))]
-                        [_                                    (loop vs (cons v real) imag)])]))]
-    [_ (values u 0)]))
-
+       [(list vs-r ... (Imaginary v-i))
```

Comment:
I don't think this works for expressions of the type `(+ 1 @i (* 2 @i) (* 3 @i))`.

Reply:
Do we need to support the unnormalized form?
We may need a normalize before matching it.